### PR TITLE
Properties and violation checking

### PR DESCRIPTION
**Pull request by malturki**
Adds a violation-recording cell and violation checking function, along with four violation properties inspired by the lucash attacks.

### CODE REVIEW

File: kmcd.md [hunk 1]
```diff
@@ -80,6 +80,165 @@ State Storage/Revert Semantics
...
+
+    rule unAuthFlapKick(.List) => false
+```
+
+### Earning interest from a pot after End is deactivated (inspired by the lucash-pot-end attack)
+
+The property checks if an `End . cage` is eventually followed by a successful `Pot . join`and `Pot . exit`.
+
+```k
+    syntax Bool ::= potEndInterest(List) [function, functional]
```
Comment (ehildenb) [inline kmcd.md]: After a call to `End . cage`, you don't see a successful call to `Pot . file dsr`

File: kmcd.md [hunk 2]
```diff
@@ -80,6 +80,165 @@ State Storage/Revert Semantics
...
+
+    rule potEndInterestEnd(.List, _) => false
+```
+
+### Earning interest from a pot in zero time (inspired by the lucash-pot attack)
+
+The property checks if a sequence of `Pot . join`, `Pot . drip` and `Pot . exit` is executed in zero time.
+
+```k
+    syntax Bool ::= zeroTimePotInterest(List) [function, functional]
```
Comment (ehildenb) [inline kmcd.md]: A call to `Pot . join` which is preceded by a `TimeStep` more recently than a `Pot . drip` should not succeed.

1.  Check for calls to Pot . drip, if so go to 1. Check for calls to TimeStep, if so go to 2.
2.  Check for calls to Pot . join, if so violation. Check for calls to Pot . drip, if so go to 1.

File: kmcd.md (hunk 1 above)
Comment (malturki) [reply kmcd.md]: Done

File: kmcd.md (hunk 2 above)
Comment (malturki) [reply kmcd.md]: Done

File: kmcd.md [hunk 3]
```diff
@@ -72,6 +72,145 @@ State Storage/Revert Semantics
...
+    rule potEndInterestEnd(.List) => false
+```
+
+### Earning interest from a pot in zero time (inspired by the lucash-pot attack)
+
+The property checks if a successful `Pot . join` is preceded by a `TimeStep` more recently than a `Pot . drip'. 
+
+```k
+    syntax Bool ::= zeroTimePotInterest(List) [function, functional]
+ // ----------------------------------------------------------------
```
Comment (ehildenb) [inline kmcd.md]: This will accept trace `TimeStep ; Pot . join ; Pot . drip`, which is an unacceptable trace.

Comment (malturki) [reply kmcd.md]: Right. This is now fixed.  I don’t think though that we will need to have the `Pot. drip` at the end for the detection pattern. `TimeStep ; Pot . join` is bad enough: once an attacker joins while `now > rho`, the damage is done.

**Pull request merged**